### DISCUSSION TITLE
2024-05-10 - VS Code C# Dev Kit - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by piers-sinclair** [SSWConsulting/SSW.Rules.Content#8530] [Chewing the Fat]
VS Code now has the C# Dev Kit.

Check out the video (6 min): 
https://www.youtube.com/watch?v=3ST-7TP09qk

What are your thoughts on a switcheroo to VS Code?

Are you dying on the Visual Studio or Rider hill?

What's the 1 feature you can't live without in your favourite IDE?

**Comment by tiagov8** [comment]
I'm not a .NET dev but I always had the feeling Visual Studio is too "heavy" and VS Code felt more lean/simple. I believe the Dev Kit creates a in between scenario.

**Comment by tkapa** [comment]
Visual Studio has always felt heavy, and it's great that there are alternatives, even though I'm not really used to them.

VSCode has been my go to for JS dev, and maybe I could find room to add C# to that list as well.

I haven't used VSCode to stand up a whole .NET application, but I might give it a go.

**Comment by GordonBeeming** [comment]
I currently switch between all the IDEs (VS, Rider and VS Code) depending on the task. Most of the time I have either VS or Rider open and VS Code at the same time as different tasks are better/quick in VS Code vs a fat client.

Co-pilot... but that is now in each, so not really something I'd need to live without

**Reply by piers-sinclair** [reply]
I switch between VS Code and Visual Studio...One day I hope to only use VS Code...but today is not that day.

**Reply by bradystroud** [reply]
@piers-sinclair profiles in VS Code could play a big part in this.
https://code.visualstudio.com/docs/editor/profiles

**Comment by brydeno** [comment]
Like Gordon I switch between them all. I'm very much a GUI dev when coding. When it comes to build and deploy, I'm in the command line / script school of thought.

**Comment by drwharris** [comment]
I also switch between VS, VSCode and now Rider. I "can" C# dev in VSCode but its not my preferred. I am so used to VS that its my preferred but of late has become slow and frustrating to work with. I have mad ethe move to rider which is much faster but I am struggling to figure out how to do things that just come naturally to me in VS.

**Comment by SamBWagner** [comment]
I think Visual Studio is perfectly fine. However, of the three, Rider is the most complete and best solution for development in .NET. That said though, VSCode would be me second choice before I chose Visual Studio as it's much lighter and, with it's extension ecosystem (which is far better than Visual Studio's) I'd put VSCode and Visual Studio on the same level.

I've also done some .NET development in VIM and I can say that it's possible and does feel nice when you just need to "Get it done". However, due to the LSP's limitations and the .NET developer's reluctance to keep it in parity with the IDE's, it can feel like you're "behind" everything else. Perfect example of this is that collection expressions (using the `[]` instead of `new List()`) still isn't supported by either Omnisharp or cs_Lang.

**Reply by BrookJeynes** [reply]
I'll agree with everything here. Unfortunately Omnisharp is not kept up to date and can lack behind a little with the newer features. This isn't a problem if you know what you're doing but you just can't beat riders optimization hints.

Vim bindings in Rider it is for now :cry:

**Comment by leahy268** [comment]
Very rarely touched visual studio as I rarely have a need. Use vscode a lot and it's quick and simple to use.  But I'm not a c# dev.

**Comment by AlexRothwell** [comment]
I've been using Rider for a while now after finding the tool offerings in resharper a great complement to an IDE. Unfortunately when compared to VS Code and Rider, the full Visual Studio just feels very slow

**Reply by jernejk** [reply]
Have you tried to uninstall R# in VS? It's usually what makes it. (and if you have Rider installed, they constantly try to re-enable/re-install R# to VS 😔)

**Comment by bradystroud** [comment]
I use Visual Studio for Mac every day, it's my favorite.

**Reply by Marxoz** [reply]
![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/86134557/b861b683-0fb3-4cd8-986b-244cc07026a2)

**Comment by AntPolkanov** [comment]
Rider - one love ❤️‍🔥

**Comment by Tanya-Leahy** [comment]
I am not a dev, so it is difficult for me to answer any questions on this.  I have just started using VS Code for editing Markdown and I am finding it simple to use.

**Comment by jakebayliss** [comment]
Usually stick to Visual Studio for .NET, anything else VSCode. Used to it and it works just fine

**Comment by matt-goldman-ssw** [comment]
Visual Studio is hands down the best dev env for .NET. Rider is good too, and it has things I prefer to VS, but on balance VS wins. I was secretly hoping that when they announced VS for Mac retirement they were going to x-plat VS, but it looks like VS Code is Microsoft's x-plat story. Which is a shame - it's good for some things, and is my favourite way to write all code that is not .NET. But it's got a long way to go to be a viable fulltime developer environment for .NET for most people.

**Comment by YalalCakirSSW** [comment]
I like working on VS especially for .NET, I've never used Rider but it looks like it's the favorite IDE so I will try it.

**Comment by Ben0189** [comment]
When working with JavaScript, I use VS Code exclusively, have some experience using VS code to do C# development but getting IntelliSense to work properly can be tricky. Still prefer using VS when working on any .NET application for now, but i am considering Rider next.

**Comment by joshbermanssw** [comment]
I use VS code when working locally on my mac and VS when on a VM - I can second the IntelliSense issues that Ben N mentioned above with C# dev in VS code. 

Keen to try Rider

**Comment by stephanfakossw** [comment]
I'm not a dev so I'm not really understanding this in full... However, I can see how this can be a useful tool for devs....

**Comment by TobyChurches** [comment]
I generally stick with VS for .NET and VS Code for frontend dev. However, this is just because it is what I am familiar + I like knowing that I can ask @jernejk if I have any VS issues 😂 (Thanks mate)

Saying that, I did like IntelliJ when I was developing in Java throughout Uni, and would be keen to give Rider a spin at some point

**Comment by Dhruv-0987** [comment]
I like the dev kit in VS code, I mostly use Rider for .NET but VS code does somethings better so its always good to have that option.

**Comment by Hona** [comment]
VS Code is a long way off being ready to compete with Rider/VS.
Rider is on top for me, and will be so for at LEAST another year or 2.

**Comment by MarcusIrmscherSSW** [comment]
I now now what an IDE is. 😅

**Comment by ricksu978** [comment]
I use VS for C#, VS Code for yml, dockerfile ...etc

C# Dev Kit on VS Code wasn't a good experience for me, seeing extension errors from time to time.
Maybe it's me not using it properly.

Must IDE Feature: F5

**Comment by zacharykeeping** [comment]
It's nice to see Microsoft continue to work on the C# experience for VS Code - I like the idea of it becoming a sort of do-it-all text editor and IDE. But it just isn't as nice an experience as a full blown IDE like VS and Rider. For me, Rider is king on my Mac.

**Comment by steven0x51** [comment]
I mainly use Visual Studio for .Net and VS Code for other things. VS Code is my favourite code editor.

**Comment by duanxinhuan** [comment]
Eventually, we have a video for VS code.
I use vs code and only vs code for everything, it's fun enjoyable and amazing.
![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/48227836/3b2c5fc3-b67e-4631-98a1-6841bdb96855)

**Comment by Calinator444** [comment]
Don't need to swtich to VS code, I already use it.

This might be a bit of laziness on my part but I mainly prefer VS code because you don't need re-learn how to do anything. I've gotten pretty good at using it. I know how to use the command palette, the source control window and I've got GitHub copilot integration set up for if I need to write something quickly.

You can also do pretty much anything in VS Code. For instance I've been using it to write python scripts and all I really needed to do was install the python package to get syntax highlighting.

The main feature I can't live without in a Code editor is syntax highlighting

**Comment by wicksipedia** [comment]
> Are you dying on the Visual Studio or Rider hill?

I chop and change between VSCode and Rider - I sometimes have the solution open in both at the same time
I like that VSCode is lightweight but some Rider features are killer e.g. refactorings, package vulnerability, code exploration

> What's the 1 feature you can't live without in your favourite IDE?

Multi-caret editing

**Comment by KostaMadorsky** [comment]
Visual Studio + ReSharper is still my go-to for .NET development.
But I use VS Code a lot with huge codebases for searching, find-and-replace with regex, etc.

**Comment by amankumarrr** [comment]
Visual Studio Code is my favorite IDE overall, but when it comes to .NET development, I always lean towards Visual Studio. It offers more IntelliSense flexibility and a broader range of debugging options.

**Comment by AttackOnMorty** [comment]
I prefer Rider for .NET development because it has more shortcuts. 

Combined with [Key Promoter X](https://plugins.jetbrains.com/plugin/9792-key-promoter-x), you can move away from the trackpad or mouse.

**Comment by jeoffreyfischer** [comment]
Great video! But I still prefer using VS for .NET project because the debugging in VSC is not as good as in VS - especially setting and using breakpoints.

**Comment by chrisschultzssw** [comment]
Not a dev so I never do C#... but I like VS Code for bicep, etc.

**Comment by tomek-i** [comment]
I love VS Code 💘 especially all those fun and useless extensions 😂 just not for .NET development.

wanna unleash your true power while coding? VS Code got your back with [Power Mode](https://marketplace.visualstudio.com/items?itemName=hoovercj.vscode-power-mode)
wanna have a pet but have no time taking care of it as you're too busy writing the next unicorn app? VS Code got you with a selection of [Pets](https://marketplace.visualstudio.com/items?itemName=tonybaloney.vscode-pets)
are inseure about your code and require constant validation ? VS Code got your back with some juicy [Rainbow Farts](https://marketplace.visualstudio.com/items?itemName=saekiraku.rainbow-fart)
still vibing the '84 synth wave? Yeaah [SynthWave '84](https://marketplace.visualstudio.com/items?itemName=RobbOwen.synthwave-vscode)
feeling nostalgic? Our most beloved [Hot Dog Stand
](https://marketplace.visualstudio.com/items?itemName=somekittens.hot-dog-stand)


I love Visual Studio for .NET development due to it's debugging tools that are lacking in VS Code like the Profiler.
Most projects I have been working on were legacy projects running on old .NET frameworks, i'd prefer running .NET 8 projects with VS Code.

Remember, VS Code is just a fancy _notepad.exe_ while [Visual Studio](https://visualstudio.microsoft.com/downloads/) is _an actual **IDE**_

**Comment by bahjatalaadelssw** [comment]
Any improvements to make workflow more organic is good. Any app that can run on MacOS is good.

**Comment by bettybondoc** [comment]
I'm not a .NET dev. VS Code is my go to IDE but I haven't really tried Visual Studio in depth to form an opinion.

**Comment by awaltos** [comment]
I love VS Code for everything at the moment, although have not used it for C# yet. I only have experience with VS for C# and found it to be ok - perhaps feeling slightly 'heavy'

> What are your thoughts on a switcheroo to VS Code?

 I would try it out on my next C# proj and if there are too many limitations there then I would next try Rider.

> Are you dying on the Visual Studio or Rider hill?  

Nope

> What's the 1 feature you can't live without in your favourite IDE?

cmd+clicking through symbols to find references/implementations

**Comment by piers-sinclair** [comment]
VS Code is awesome, specifically because it fully supports dev containers, unlike Visual Studio.

**Comment by christoment** [comment]
> What are your thoughts on a switcheroo to VS Code?

VSCode is good and lightweight, but it is not an IDE - it can do .NET development to some extend, but you will miss some experience which only exists in full blown IDE.

I think when you just want to explore the code and do some lightweight code editing, VSCode is perfect, it has intellisense that will assist you just enough. 😄

> Are you dying on the Visual Studio or Rider hill?

Yes, using VisualStudio feels like walking underwater in a cave - it's so slow, sluggish, and claustrophobic 😅
This is what it feels like working in VisualStudio:
![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/35278147/cc4494e9-968c-4aa3-9b2c-c2a0d28724aa)

Rider is much faster and the UI feels more modern and less hostile.

The only downside to Rider is that new .NET support won't immediately available in Rider - only Visual Studio will support latest .NET features immediately on launch.

**Comment by JeanThirion** [comment]
> What are your thoughts on a switcheroo to VS Code?

A what sorry ? Haha.
Looks like they're trying hard - and I can't see why they wouldn't succeed in making VSCode better than VS ever was. It's actually probably bound to happen. 
I don't think I'll transition just yet, but seems like I ultimately will

> Are you dying on the Visual Studio or Rider hill?

I don't die.

> What's the 1 feature you can't live without in your favourite IDE?

Pets !
https://marketplace.visualstudio.com/items?itemName=tonybaloney.vscode-pets

**Comment by Marxoz** [comment]
Rider all the way for me. I find it easier to navigate, and definitely a lot faster. I am on a Mac though. If I was on Windows I won't mind using VS, but VS for Mac is a disaster.

**Comment by cvionnet** [comment]
@AntPolkanov  will burn me alive, but every time I tried using Rider extensively, I ended up coming back to Visual Studio.
One missing feature in Rider is the ability to group files by project:

![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/46954368/e43a0955-0a4c-435a-9a61-063b65f3ac8d)
Figure: Tabs Explorer in VS showing files grouped by project

If anyone knows how to do it, please share!

**Comment by ozair-ashfaque** [comment]
I would prefer Visual Studio for .Net Development and VS Code for Angular or React etc.

**Comment by MattParkerDev** [comment]
VS Code has a long way to go before it can even compete with Visual Studio, let alone Rider.

**Comment by JackDevAU** [comment]
🆗 VS is great - has a very powerful debugger and an extensive amount of dev tools but Rider is just better.
✅ Rider - has great intellisense and awesome refactoring. My current go to for coding with C#
😢 VS Code - lightweight, great for quick edits... will have to check out how it goes with Dev Kit

**Comment by jerryluossw** [comment]
if I'm using windows, I would prefer Visual Studio for .Net Development and VS Code for Other languages, Such as front-end languages, Python and so on.
I think Solution Explorer of C# Dev Kit is great, as it becomes more and more powerful, I will try to use it more on small projects, because it's very lightweight.

**Comment by PennyWalker** [comment]
I'm not a dev, so I'm prob not as qualified to comment but adding anything that makes it easier to use, like this tool kit will benefit people who are just getting started and who need more help using such a monolith (and I'm referring to ancient monoliths here, not software architecture :D).

From my conversations with the devs, people use the different options interchangeably, but Rider is still the preferred choice for a lot of work - it's more user-friendly.

**Comment by SofieHong** [comment]
This video is well-produced. The editing pace and transitions between different scenes are excellent!

**Comment by jernejk** [comment]
VS Core, Visual Studio and Rider have pros and cons, so I'm using them all.

Visual Studio:
✅ Awesome for C#
✅ Awesome for the majority of reactors
✅ Awesome git features (they work for me, even git merge)
✅ Live unit tests
✅ Highly predictable typing experience. IntelliSense is bound by key presses rather than async. Even if VS freezes for 5 seconds, I can still type and leverage my muscle memories to predict what IntelliSense will offer and select it even I won't see any of the action happening for 5 seconds (I worked with VS for about 15 years 😅)
✅ Real fast 80% of the time
✅ Get the latest features for current and preview .NET
❌ Can take a lot of RAM
❌ When it's slow, it's really slow
❌ Can occasionally some weird issues
❌ Blazor support is just OK

Rider (not actively using atm!):
✅ Awesome for Blazor
✅ Awesome for performance analysis
✅ Awesome for Macs
✅ Has some unique refactoring capabilities
✅ Fast IDE
❌ Has some odd issues that might conflict with Visual Studio and VS Code (like project GUID in sln files)
❌ By default, uses Mono coding standards instead of official C# coding standards (requires large `.editorconfig` for no good reason)
❌ Due to async IntelliSense (as opposed key-bound IntelliSense), I can't type faster than IntelliSense, I need to wait for it
❌ I don't particularly appreciate how they run "live" unit tests

VS Code:
✅ Awesome for UI code like JavaScript and TypeScript
✅ Awesome for Markdown
✅ Awesome for Mermaid
✅ Awesome for extensions
✅ Awesome for presentations and demos
✅ Awesome for new people
✅ Awesome for plain text and analyzing text data
✅ Great for most scripting languages and DevOps
✅ Can run in the browser and even in code spaces
✅ Has a lot of extensions!
❌ Don't like it for C#
❌ Blazor doesn't seem to work much better than VS
❌ Debugging can be a pain
❌ Might require a lot of extensions!
❌ Managing extension per workspace can be a pain

**Comment by Aibono1225** [comment]
I prefer Visual Studio for .NET development, and VS Code for frontend projects.
But I would like to try out C# Dev Kit

**Comment by jimmidier** [comment]
I have switched to VS Code for full stack development for almost a year now and it's a really good experience for the consistency. Even back when the C# extension was still using Omnisharp it worked seamlessly and there've really been no obvious downside comparing to Visual Studio.

**Comment by tino-liu** [comment]
I've been using VS Code and Rider at the same time. I love fast it is to develop on the Mac with both of them. Personally I think VS Code is king in terms of versitility.

**Comment by KristenHu** [comment]
I'd like to use visual studio for .NET development, but I'll give it a try to use VS code next time.

**Comment by babakamyljanovssw** [comment]
I use VS for .NET dev, and havent't tried yet Rider and VS Code for .NET dev. I use VS Code for Angular development daily.
The only pain I'm facing with VS is it takes a lot of RAM, and can be slow sometimes.

**Comment by tombui99** [comment]
As a Mac user, I vouch for VS Code.

**Comment by PothieuG** [comment]
I tried VS Code for the .NET environment a few years ago, and it was awful... Maybe I should give it another try today.
For the time being, I prefer to work with Visual Studio for the .NET environment, and use VS Code for everything else (mostly frontend development).

**Comment by danielmackay** [comment]
> What are your thoughts on a switcheroo to VS Code?

No thanks.

> Are you dying on the Visual Studio or Rider hill?

I'm not dying on any hills, but I really like Rider and don't plan on switching from it anytime soon.

> What's the 1 feature you can't live without in your favourite IDE?

All the awesome refactoring that comes with Rider or Resharper.

**Comment by vladislav-kir** [comment]
I usually use both Visual Studio and VS Code. First for .NET and second for everything else. 

I am satisfied to how that works, but interested to see if I could move to VS Code completely.
